### PR TITLE
Implement context logger for remaining packages

### PR DESCRIPTION
**Pull request by klingtnet**
This is a follow up to #93 and implements the context-logger for the remaining packages.
~~Please _don't merge this right now_ because I want to give this a review tomorrow.~~

### CODE REVIEW

**Comment by heyLu** [comment]
Logging in [cbreaker/predicates.go:53](https://github.com/vulcand/oxy/blob/master/cbreaker/predicates.go#L53) seems possible, as it returns a closure which has access to the `CircuitBreaker` struct.

File: buffer/buffer.go [hunk 1]
```diff
@@ -99,6 +103,16 @@ func New(next http.Handler, setters ...optSetter) (*Buffer, error) {
 	return strm, nil
 }
 
+// Logger defines the logger the forwarder will use.
```
Comment (heyLu) [inline buffer/buffer.go]: Stray mention of "forwarder" here, should be "buffer" or something similar.

File: cbreaker/cbreaker.go [hunk 2]
```diff
@@ -99,9 +102,19 @@ func New(next http.Handler, expression string, options ...CircuitBreakerOption)
 	return cb, nil
 }
 
+// Logger defines the logger the forwarder will use.
```
Comment (heyLu) [inline cbreaker/cbreaker.go]: Another stray mention of "forwarder" in the doc comment.

**Comment by klingtnet** [comment]
@heyLu Right, I have just fixed this.

File: connlimit/connlimit.go [hunk 3]
```diff
@@ -41,24 +43,36 @@ func New(next http.Handler, extract utils.SourceExtractor, maxConnections int64,
...
 	if cl.errHandler == nil {
-		cl.errHandler = defaultErrHandler
+		cl.errHandler = &ConnErrHandler{
+			log: cl.log,
+		}
 	}
 	return cl, nil
 }
 
+// Logger defines the logger the forwarder will use.
```
Comment (heyLu) [inline connlimit/connlimit.go]: Another stray mention of "forwarder" in the doc comment.

File: ratelimit/tokenlimiter.go [hunk 4]
```diff
@@ -98,6 +102,16 @@ func New(next http.Handler, extract utils.SourceExtractor, defaultRates *RateSet
 	return tl, nil
 }
 
+// Logger defines the logger the forwarder will use.
```
Comment (heyLu) [inline ratelimit/tokenlimiter.go]: Another stray mention of "forwarder" in the doc comment.

File: roundrobin/rebalancer.go [hunk 5]
```diff
@@ -123,6 +127,16 @@ func NewRebalancer(handler balancerHandler, opts ...RebalancerOption) (*Rebalanc
 	return rb, nil
 }
 
+// Logger defines the logger the forwarder will use.
```
Comment (heyLu) [inline roundrobin/rebalancer.go]: Another stray mention of "forwarder" in the doc comment.

File: roundrobin/rr.go [hunk 6]
```diff
@@ -67,13 +71,23 @@ func New(next http.Handler, opts ...LBOption) (*RoundRobin, error) {
 	return rr, nil
 }
 
+// Logger defines the logger the forwarder will use.
```
Comment (heyLu) [inline roundrobin/rr.go]: Another stray mention of "forwarder" in the doc comment.

File: stream/stream.go [hunk 7]
```diff
@@ -73,6 +77,16 @@ func New(next http.Handler, setters ...optSetter) (*Stream, error) {
 	return strm, nil
 }
 
+// Logger defines the logger the forwarder will use.
```
Comment (heyLu) [inline stream/stream.go]: Another stray mention of "forwarder" in the doc comment.

File: stream/stream_test.go [hunk 8]
```diff
@@ -326,21 +327,23 @@ func (s *STSuite) TestPreservesTLS(c *C) {
 }
 
 func BenchmarkLoggingDebugLevel(b *testing.B) {
-	streamer, _ := New(noOpNextHttpHandler{})
+	l := log.StandardLogger()
+	streamer, _ := New(noOpNextHttpHandler{}, Logger(l))
```
Comment (heyLu) [inline stream/stream_test.go]: I think the changes in this benchmark are unnecessary, as the default is logrus' StandardLogger, and `SetLevel` und `SetOutput` work on that logger.

File: stream/stream_test.go [hunk 9]
```diff
@@ -326,21 +327,23 @@ func (s *STSuite) TestPreservesTLS(c *C) {
...
 
 	for i := 0; i < b.N; i++ {
 		heavyServeHttpLoad(streamer)
 	}
 }
 
 func BenchmarkLoggingInfoLevel(b *testing.B) {
-	streamer, _ := New(noOpNextHttpHandler{})
+	l := log.StandardLogger()
+	streamer, _ := New(noOpNextHttpHandler{}, Logger(l))
```
Comment (heyLu) [inline stream/stream_test.go]: Similarly unnecessary changes as above.

File: trace/trace.go [hunk 10]
```diff
@@ -69,14 +73,24 @@ func New(next http.Handler, writer io.Writer, opts ...Option) (*Tracer, error) {
 	return t, nil
 }
 
+// Logger defines the logger the forwarder will use.
```
Comment (heyLu) [inline trace/trace.go]: Another stray mention of "forwarder" in the doc comment.

File: utils/handler.go [hunk 11]
```diff
@@ -4,15 +4,20 @@ import (
...
 )
 
 type ErrorHandler interface {
 	ServeHTTP(w http.ResponseWriter, req *http.Request, err error)
 }
 
-var DefaultHandler ErrorHandler = &StdHandler{}
+var DefaultHandler ErrorHandler = &StdHandler{
+	log: log.StandardLogger(),
+}
```
Comment (heyLu) [inline utils/handler.go]: Is this logger used anywhere?  I couldn't find any usages.  (Except in this file, but no calls to any `log.*` function.)

File: buffer/buffer.go (hunk 1 above)
Comment (klingtnet) [reply buffer/buffer.go]: Ah, yeah copy-paste errors.

**Comment by klingtnet** [comment]
Ok, the MR is now ready to merge, at least from my side.
Should I squash the commits?

**Comment by klingtnet** [comment]
Are there any plans to merge this?

**Comment by archisgore** [comment]
Sorry about that. I just resolved conflicts. If the build passes, I'll merge this in.
